### ISSUE TITLE
Some apparent duplication: binary naturals vs binary positive naturals

### ISSUE THREAD

**Issue by abooij** [HoTT/HoTT#1263]
#1137 defined a type of positive naturals in [`theories/Spaces/Pos/Core.v`](https://github.com/HoTT/HoTT/blob/7910fb7fe6390320438608a5480b5947f66d4eb6/theories/Spaces/Pos/Core.v#L8):
```coq
Inductive Pos : Type0 :=
  | xH : Pos
  | x0 : Pos -> Pos
  | x1 : Pos -> Pos.
```
There exists a very similar development of binary naturals in [`theories/Classes/implementations/binary_naturals.v`](https://github.com/HoTT/HoTT/blob/7910fb7fe6390320438608a5480b5947f66d4eb6/theories/Classes/implementations/binary_naturals.v#L23), merged in #919:
```coq
Inductive binnat : Type :=
  | bzero   :           binnat  (* zero *)
  | double1 : binnat -> binnat  (* 2n+1 *)
  | double2 : binnat -> binnat. (* 2n+2 *)
```

Both parts of the library develop successor functions, addition, multiplication, associativity laws, etc. Is there a straightforward way to deduplicate these developments? I believe the main application of the former development is in `theories/Spaces/Int/Core.v`, right?

**Comment by Alizter** [comment]
So here is a silly answer: the positives don't have a zero.

I wrote the binary positives based off of the coq stdlib implementation.
This allowed us to have an efficient definition of the integers. Previously
it was two copies of unary positives(/nats) stuck together with a zero.

I am currently working on better integrating the  hottclasses stuff with
the rest if the library. The original math classes library wasn't designed
with HoTT in mind so what we have currently is an adaptation. Not a bad
one, but one that needs some work.

I don’t currently think that this is a code duplication that is worth
eliminating. Since positives are printed/notated in a particular way that
allows working with integers to be easier.

I would highly not recommend using Pos for anything other than the
integers. Think of them as a helper type for the binary integers.

**Comment by mikeshulman** [comment]
It does seem that the binary positives are more useful for defining binary integers than the binary naturals are.  But it should also be easy to define a notion of binary naturals in terms of the binary positives, by adding zero but no negatives, and it might be easy to deduce all the operations and properties of these binary naturals from those of the binary positives.  That might reduce some duplication.

**Comment by jdchristensen** [comment]
Jarl and I have been finding the Integers defined using the binary naturals to be pretty painful to deal with.  You have to case split on pos, neg and zero, and for the pos and neg cases, not very much is known about the binary positives.  If instead things were defined with the usual natural numbers (so we'd have constructors pos, neg : nat -> Int (and zero : Int)), then we'd be able to make use of the usual properties.  This would of course be less efficient, but for our library it seems more important to be able to prove things than compute with them.

(I also find the near duplication of code from math classes annoying.  I feel that we should reduce our dependence on the math classes code when possible.)

**Comment by jdchristensen** [comment]
I see that there is a related issue: #1154.  (And I forgot to tag @jarlg .)  Maybe efficiency is more important than I realize?

In a slightly different direction, is there a very complete library about the integers that we can just import?  It seems silly for us to reinvent the wheel.  As a test, it should contain a proof that multiplication by a non-zero integer gives an injection from the integers to itself...

**Comment by mikeshulman** [comment]
Back in #1124 we considered using the (binary) integers from the Coq stdlib.

**Comment by jdchristensen** [comment]
Thanks, that explains why it's not so easy to reuse code from the stdlib.  That's unfortunate.

I'm still unsure about whether working with binary numbers is worth it.  I guess an ideal situation would be to have multiple implementations which are shown to be equivalent, and then to prove properties using the implementation that makes that easiest.  I tried starting with the free group on one generator as the definition.  It's slightly non-trivial from just the definition to show that it is abelian, but I did so.  But showing that multiplication by a non-zero element is an injection seems tricky, just using the universal property.

**Comment by Alizter** [comment]
@jdchristensen Do you find working with binary positives difficult? There are some custom induction principles for doing case analysis on successors. However there is very little arithmetic for Pos (but also nat too).

**Comment by jdchristensen** [comment]
@Alizter Yes, I found working with binary positives difficult.  Too many cases arise.  Even with the custom induction principle, you end up having to reprove things for Pos that we already know are true for nat, which is annoying.  Proving the obvious bijection between Pos and nat would help.
